core/src/main/scala/cats/Monad.scala

```diff
@@ -14,4 +14,86 @@ import simulacrum.typeclass
 @typeclass trait Monad[F[_]] extends FlatMap[F] with Applicative[F] {
   override def map[A, B](fa: F[A])(f: A => B): F[B] =
     flatMap(fa)(a => pure(f(a)))
+
+  /**
+   * Execute an action repeatedly as long as the given `Boolean` expression
+   * returns `true`. The condition is evaluated before the loop body.
+   * Collects the results into an arbitrary `Alternative` value, such as a `Vector`.
+  */
+  def whileM[G[_], A](p: F[Boolean])(body: => F[A])(implicit G: Alternative[G]): F[G[A]] = {
+    val b = Eval.later(body)
+    tailRecM[G[A], G[A]](G.empty)(xs => ifM(p)(
+      ifTrue = {
+        map(b.value) { bv =>
+          Left(G.combineK(xs, G.pure(bv)))
```

Comment:
Notice here that to maintain the order of evaluation results I have to append the latest result to the end of the `Alternative`, which means the performance will be horrible for `List`. 
I see 2 options here: 
1. in the API document, point this out and warn people against data structures with non-constant append performance. 
2. use prepend here and return the `G` with results in the reverse order, i.e. last result first. 
3. maintain an internal `List` in the reverse order and construct the `G` at last in the right order. 

@tpolecat and @Atry, what would be your preference here?